kategory-docs/docs/docs/patterns/monadcomprehensions/README.md

```diff
@@ -6,4 +6,211 @@ permalink: /docs/patterns/monad_comprehensions/
+`yields()` is a helper function that takes any one value and constructs a correct return.
+Let's see a minimal example.
+
+```kotlin:ank
+import kategory.*
+import kategory.effects.*
+
+IO.monad().binding {
+  yields(1)
+}.ev().unsafeRunSync()
+```
+
+Anything in the function inside `binding` can be imperative and sequential code that'll be executed when the datatype decides.
+In the case of [`IO`]({{ '/docs/effects/io' | relative_url }}), it is immediately run blocking the current thread using `unsafeRunSync()`. Let's expand the example by adding a second operation:
+
+```kotlin
+IO.monad().binding {
+  val a = IO.invoke { 1 }
+  yields(a + 1)
+}.ev().unsafeRunSync()
+// Compiler error: the type of a is IO<Int>, cannot add 1 to it
+```
+
+This is our first challenge. We've created an instance of [`IO`]({{ '/docs/effects/io' | relative_url }}) that'll run a block asynchronously, and we cannot get the value from inside it.
+From the previous snippet the first intuition would be to call `unsafeRunSync()` on `a` to get the value.
+This will block the current thread until the operation completes. What we want is to, instead, run and await until `a` completes before yielding the result.
+For that we have two flavors of the function `bind()`, which is a function only available inside the function passed to `binding()`.
+
+```kotlin:ank
+IO.monad().binding {
+  val a = IO.invoke { 1 }.bind()
+  yields(a + 1)
+}.ev().unsafeRunSync()
+```
+
+```kotlin:ank
+IO.monad().binding {
+  val a = bind { IO.invoke { 1 } }
+  yields(a + 1)
+}.ev().unsafeRunSync()
+```
+
+What `bind()` does is use the rest of the sequential operations as the function you'd normally past to `flatMap`.
+The equivalent code without using comprehensions would look like:
+
+```kotlin:ank
+IO.invoke { 1 }
+  .flatMap { result ->
+    IO.pure(result + 1)
+  }
+.ev().unsafeRunSync()
+```
+
+With this new style we can rewrite our original example of database fetching as:
+
+```kotlin
+val university: IO<University> = 
+  IO.monad().binding {
+    val student = getStudentFromDatabase("Bob Roxx").bind()
+    val university = getUniversityFromDatabase(student.universityId).bind()
+    val dean = getDeanFromDatabase(university.deanId).bind()
+    yields(dean)
+  }
+```
+
+And you can still write your usual imperative code in the binding block, interleaved with code that returns instances of [`IO`]({{ '/docs/effects/io' | relative_url }}).
+
+```kotlin
+fun getNLines(path: FilePath, count: Int): IO<List<String>> = 
+  IO.monad().binding {
+    val file = getFile(path).bind()
+    val lines = file.readLines().bind()
+    if (lines.length < count) {
+      IO.raiseError(RuntimeException("File has fewer lines than expected"))
+    } else {
+      yields(lines.take(count))
+    }
+  }
+```
+
+While this looks like a great improvement to manually raise errors sometimes you will encounter unexpected behavior and exceptions in seemingly normal code.
+
+### Error handling in comprehensions
```

Comment:
Perhaps link to `docs/patterns/error_handling`?

Reply:
I should rename it, this is about wrapping arbitrary errors in the chain.